rhizome/lib/vm_setup.rb

```diff
@@ -265,24 +282,121 @@ def setup_volume(storage_volume, index, boot_image)
     vp.vhost_sock(index)
   end
 
-  def setup_disk_file(storage_volume, index, boot_image)
-    disk_file = vp.disk(index)
+  def setup_spdk_bdev(bdev, disk_file, encrypted, encryption_key)
+    q_bdev = bdev.shellescape
     q_disk_file = disk_file.shellescape
 
-    if storage_volume["boot"]
-      image_path = download_boot_image(boot_image)
+    if encrypted
+      q_keyname = "#{bdev}_key".shellescape
+      q_aio_bdev = "#{bdev}_aio".shellescape
+      r "#{Spdk.rpc_py} accel_crypto_key_create " \
+        "-c #{encryption_key[:cipher].shellescape} " \
+        "-k #{encryption_key[:key].shellescape} " \
+        "-e #{encryption_key[:key2].shellescape} " \
+        "-n #{q_keyname}"
+      r "#{Spdk.rpc_py} bdev_aio_create #{q_disk_file} #{q_aio_bdev} 512"
+      r "#{Spdk.rpc_py} bdev_crypto_create -n #{q_keyname} #{q_aio_bdev} #{q_bdev}"
+    else
+      r "#{Spdk.rpc_py} bdev_aio_create #{q_disk_file} #{q_bdev} 512"
+    end
+  end
 
-      # Images are presumed to be atomically renamed into the path,
-      # i.e. no partial images will be passed to qemu-image.
-      r "qemu-img convert -p -f qcow2 -O raw #{image_path.shellescape} #{q_disk_file}"
+  def setup_data_encryption_key(index, key_wrapping_secrets)
+    data_encryption_key = OpenSSL::Cipher.new("aes-256-xts").random_key.unpack1("H*")
+
+    result = {
+      cipher: "AES_XTS",
+      key: data_encryption_key[..63],
+      key2: data_encryption_key[64..]
+    }
+
+    key_file = vp.data_encryption_key(index)
+
+    # save encrypted key
+    sek = StorageKeyEncryption.new(key_wrapping_secrets)
+    sek.write_encrypted_dek(key_file, result)
+
+    FileUtils.chown @vm_name, @vm_name, key_file
+    FileUtils.chmod "u=rw,g=,o=", key_file
+
+    r "sync #{key_file.shellescape}"
+
+    result
+  end
+
+  def copy_image(disk_file, boot_image, disk_size_gib, encrypted, encryption_key)
+    image_path = download_boot_image(boot_image)
+
+    size = File.size(image_path)
+
+    fail "Image size greater than requested disk size" unless size <= disk_size_gib * 2**30
+
+    bdev_conf = [{
+      method: "bdev_aio_create",
+      params: {
+        name: "aio0",
```

Comment:
I'm a bit puzzled by the name parameter here, what's it naming? Is it namespaced somehow? Does something bad happen if two images get involved, both represented by `aio0` varyingly, especially under concurrent access?

Reply:
This doesn't involve the main spdk vhost service & doesn't create any bdevs there. This is a single `spdk_dd` process which will have just one (if unencrypted) or two (if encrypted) bdevs & will use those bdevs to copy data in proper form into target file.

So two images are isolated by two `spdk_dd` processes & name conflict is not relevant.

Reply:
got it, nice